contracts/BoostedMasterChefJoe.sol

```diff
@@ -0,0 +1,621 @@
+        uint256 lpPercent = total.sub(MASTER_CHEF_V2.devPercent()).sub(MASTER_CHEF_V2.treasuryPercent()).sub(
+            MASTER_CHEF_V2.investorPercent()
+        );
+        uint256 lpShare = MASTER_CHEF_V2.joePerSec().mul(lpPercent).div(total);
+        amount = lpShare.mul(MASTER_CHEF_V2.poolInfo(MASTER_PID).allocPoint).div(MASTER_CHEF_V2.totalAllocPoint());
+    }
+
+    /// @notice Update reward variables of the given pool.
+    /// @param pid The index of the pool. See `poolInfo`.
+    function updatePool(uint256 pid) public {
+        PoolInfo memory pool = poolInfo[pid];
+        if (block.timestamp > pool.lastRewardTimestamp) {
+            uint256 lpSupply = pool.lpToken.balanceOf(address(this));
+            if (lpSupply > 0) {
+                uint256 secondsElapsed = block.timestamp.sub(pool.lastRewardTimestamp);
+                uint256 joeReward = secondsElapsed.mul(joePerSec()).mul(pool.allocPoint).div(totalAllocPoint);
+                pool.accJoePerShare = pool.accJoePerShare.add(
+                    (joeReward.mul(ACC_TOKEN_PRECISION).div(pool.totalBoostedAmount))
+                );
+            }
+            pool.lastRewardTimestamp = block.timestamp;
+            poolInfo[pid] = pool;
+            emit UpdatePool(pid, pool.lastRewardTimestamp, lpSupply, pool.accJoePerShare);
+        }
+    }
+
+    /// @notice Deposit LP tokens to BMCJ for JOE allocation.
+    /// @param pid The index of the pool. See `poolInfo`.
+    /// @param amount LP token amount to deposit.
+    function deposit(uint256 pid, uint256 amount) external nonReentrant {
+        harvestFromMasterChef();
+        updatePool(pid);
+        PoolInfo storage pool = poolInfo[pid];
+        UserInfo storage user = userInfo[pid][msg.sender];
+
+        if (user.amount > 0) {
+            // Harvest JOE
+            uint256 boostedAmount = user.amount.mul(getUserBoost(pid, msg.sender)).div(1000);
+
+            uint256 pending = boostedAmount.mul(pool.accJoePerShare).div(ACC_TOKEN_PRECISION).sub(user.rewardDebt).add(
+                claimableJoe[pid][msg.sender]
+            );
+            claimableJoe[pid][msg.sender] = 0;
+            JOE.safeTransfer(msg.sender, pending);
+            emit Harvest(msg.sender, pid, pending);
+        }
+
+        uint256 balanceBefore = pool.lpToken.balanceOf(address(this));
+        pool.lpToken.safeTransferFrom(msg.sender, address(this), amount);
+        uint256 receivedAmount = pool.lpToken.balanceOf(address(this)).sub(balanceBefore);
+
+        // Effects
+
+        /// If a user already has an amount we need to update the `poolInfo`
+        /// `totalBoostedAmount` and `totalVeJoe` variables relative to their
+        /// old amount. Otherwise we can just add the new amounts onto the
+        /// existing.
+
+        // Update the `pool.totalVeJoe`.
+        pool.totalVeJoe = pool.totalVeJoe.sub(user.veJoeBalance).add(VEJOE.balanceOf(msg.sender));
+        user.veJoeBalance = VEJOE.balanceOf(msg.sender);
```

Comment:
two thoughts here regarding potentially saving some gas:
1. Maybe we can check if `user.veJoeBalance != VEJOE.balanceOf(msg.sender)` and only perform the update to `pool.totalVeJoe` if that is the case
2. We can store `VEJOE.balanceOf(msg.sender)` into a local variable so that we don't have to call it twice

Reply:
Have removed this altogether as all veJoe balance changes should now be updated by the token callback function.